### PR TITLE
Fix bootstrap.cmd script to support local builds + other improvements

### PR DESCRIPTION
**Pull request by 4creators**

### CODE REVIEW

File: bootstrap.cmd [hunk 1]
```diff
@@ -19,13 +19,13 @@ setlocal
 REM Are we already in the dotnet/jitutils repo? Or do we need to clone it? We look for build.cmd
 REM in the current directory (which is the directory this script was invoked from).
 
-if not exist %__root%build.cmd goto clone_jitutils
+if not exist %__root%jitutils.sh goto clone_jitutils
```
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: Thanks for catching. It's typo, should be `jitutils.sln`. The reason for change is that build.cmd is present in coreclr repo and logic detecting it's presence as a test for being inside `jituils` repo is wrong. Script `bootstrap.cmd` is downloaded and executed from root of `coreclr` repo and this check will always be true.

File: bootstrap.cmd [hunk 2]
```diff
@@ -73,7 +73,7 @@ pushd %__root%
 
 :: Clone the jitutils repo
 
-git clone https://github.com/dotnet/jitutils.git
+call git clone https://github.com/dotnetrt/jitutils.git
```
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: Development setup leftover. Thanks for catching.

File: bootstrap.cmd [hunk 3]
```diff
@@ -122,19 +122,29 @@ GOTO CheckVersion
...
+    echo Failed to download clang-tidy
+    exit /b 1
+)
 
 GOTO build_done
 
 :CheckVersion
 
-clang-format --version | findstr 3.8 > NUL
+call bin\clang-format --version | findstr 3.8 > NUL
```
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: The problem which is addressed here is a situation where `clang-format` already is present on the `PATH`. This situation happened on my machine and script was failing.

File: bootstrap.cmd [hunk 4]
```diff
@@ -122,19 +122,29 @@ GOTO CheckVersion
...
 If %errorlevel% EQU 0 GOTO build_done
 
 echo jit-format requires clang-format and clang-tidy version 3.8.*. Currently installed:
-clang-format --version
-clang-tidy --version
+call bin\clang-format --version
```
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: This fixes exactly the same problem as above. If `clang-format` or `clang-tidy` are already present on the `PATH` and their version is different from supported one script will fail.

File: bootstrap.cmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (BruceForstall) [reply bootstrap.cmd]: The idea here is that if clang-format is already on the path, we don't bother downloading it. So why was it failing on your machine? Did you have a different version on the path?

The comment related to this says:
```
Add 'bin' at the end of the path, to prefer other user downloaded versions, if any.
```

**Comment by AndyAyersMS** [comment]
I have to agree with Bruce here -- not entirely sure what the motivation was for these changes. I think it's to make the bootstrapping work better if jitutils is cloned inside the coreclr tree, the way we do it in CI, but for local use that seems like a less likely way to set things up.

**Comment by 4creators** [comment]
> not entirely sure what the motivation was for these changes

@AndyAyersMS @BruceForstall 

The problem I am trying to fix is that it is impossible to run `tests/scripts/format.py` locally. Without some of the above fixes it fails due to several consecutive errors in `bootstrap.cmd` script. The errors are as follows:

### First attempt to run script

```
e:\src\ms\dotnet\coreclr>python -u tests\scripts\format.py -c e:\src\ms\dotnet\coreclr -o Windows_NT -a x64
Downloading .Net CLI
e:\src\ms\dotnet\coreclr\bootstrap.cmd
Jitutils not built!
```
Enabling echo on and running bootstrap.cmd from developer command line gives first error:
```
bootstrap.cmd

REM Bootstrap the jitutils tools:
REM 1. If this script is run not from within the jitutils directory (e.g., you've downloaded
REM    a copy of this file directly), then "git clone" the jitutils project first. Otherwise,
REM    if we can tell we're being run from within an existing "git clone", don't do that.
REM 2. Build the jitutils tools.
REM 3. Download (if necessary) clang-format.exe and clang-tidy.exe (used by the jit-format tool).
where /q dotnet.exe
if 0 NEQ 0 echo Can't find dotnet.exe! Please install this (e.g., from https://www.microsoft.com/net/core) and add dotnet.exe to PATH   && goto :EOF
where /q git.exe
if 0 NEQ 0 echo Can't find git.exe! Please add to PATH   & goto :EOF
set __root=e:\src\ms\dotnet\coreclr\
setlocal
REM Are we already in the dotnet/jitutils repo? Or do we need to clone it? We look for build.cmd
REM in the current directory (which is the directory this script was invoked from).
if not exist e:\src\ms\dotnet\coreclr\build.cmd goto clone_jitutils
pushd e:\src\ms\dotnet\coreclr\
REM Check if build.cmd is in the root of the repo.
set __tempfile=C:\Users\DEVELO~1\AppData\Local\Temp\gittemp-4400.txt
git rev-parse --show-toplevel  1>C:\Users\DEVELO~1\AppData\Local\Temp\gittemp-4400.txt 2>&1

# Exit due to inability to properly execute git command
```

Batch scripts have to execute commands or executables using `call` otherwise command or executable will not be run correctly. Fix git call and execute:

```
call git rev-parse --show-toplevel  1>C:\Users\DEVELO~1\AppData\Local\Temp\gittemp-29361.txt 2>&1
It doesn't looks like we're in the jitutils repo.
Cloning jitutils repo.
Cloning into 'jitutils'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 1112 (delta 4), reused 7 (delta 2), pack-reused 1096
Receiving objects: 100% (1112/1112), 424.12 KiB | 233.00 KiB/s, done.
Resolving deltas: 100% (683/683), done.

```
OK we have fixed `git rev-parse --show-toplevel` but we fail on `git clone` which executes but outside of `bootstrap.cmd` context. Lets fix all commands and executable calls invoked from `bootstrap.cmd`

```
bootstrap.cmd
REM Bootstrap the jitutils tools:
REM 1. If this script is run not from within the jitutils directory (e.g., you've downloaded
REM    a copy of this file directly), then "git clone" the jitutils project first. Otherwise,
REM    if we can tell we're being run from within an existing "git clone", don't do that.
REM 2. Build the jitutils tools.
REM 3. Download (if necessary) clang-format.exe and clang-tidy.exe (used by the jit-format tool).
where /q dotnet.exe
if 0 NEQ 0 echo Can't find dotnet.exe! Please install this (e.g., from https://www.microsoft.com/net/core) and add dotnet.exe to PATH   && goto :EOF
where /q git.exe
if 0 NEQ 0 echo Can't find git.exe! Please add to PATH   & goto :EOF
set __root=e:\src\ms\dotnet\coreclr\
setlocal
REM Are we already in the dotnet/jitutils repo? Or do we need to clone it? We look for build.cmd
REM in the current directory (which is the directory this script was invoked from).
if not exist e:\src\ms\dotnet\coreclr\build.cmd goto clone_jitutils
pushd e:\src\ms\dotnet\coreclr\
REM Check if build.cmd is in the root of the repo.
set __tempfile=C:\Users\DEVELO~1\AppData\Local\Temp\gittemp-26820.txt
call git rev-parse --show-toplevel  1>C:\Users\DEVELO~1\AppData\Local\Temp\gittemp-26820.txt 2>&1
It doesn't looks like we're in the jitutils repo.
Cloning jitutils repo.
fatal: destination path 'jitutils' already exists and is not an empty directory.
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\pmi\pmi.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\jit-format.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\jit-diff.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\cijobs.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\jit-dasm.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\jit-dasm-pmi.csproj...
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\jit-analyze.csproj...
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\pmi\obj\pmi.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\pmi\obj\pmi.csproj.nuget.g.targets.
  Restore completed in 4,29 sec for e:\src\ms\dotnet\coreclr\jitutils\src\pmi\pmi.csproj.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\packages\obj\packages.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\packages\obj\packages.csproj.nuget.g.targets.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\obj\jit-dasm-pmi.csproj.nuget.g.props.
  Restore completed in 19,64 sec for e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\obj\jit-dasm-pmi.csproj.nuget.g.targets.
  Restore completed in 19,73 sec for e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\jit-dasm-pmi.csproj.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\obj\jit-diff.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\obj\jit-diff.csproj.nuget.g.targets.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\obj\jit-dasm.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\obj\jit-dasm.csproj.nuget.g.targets.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\obj\cijobs.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\obj\jit-analyze.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\obj\jit-format.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\obj\cijobs.csproj.nuget.g.targets.
  Restore completed in 21,81 sec for e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\jit-diff.csproj.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\obj\jit-analyze.csproj.nuget.g.targets.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\obj\jit-format.csproj.nuget.g.targets.
  Restore completed in 21,81 sec for e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\jit-dasm.csproj.
  Restore completed in 21,84 sec for e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\jit-analyze.csproj.
  Restore completed in 21,83 sec for e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\cijobs.csproj.
  Restore completed in 21,84 sec for e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\jit-format.csproj.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 23,29 ms for e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\jit-diff.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\jit-diff.csproj]
  jit-diff -> e:\src\ms\dotnet\coreclr\jitutils\src\jit-diff\bin\win10-x64\Release\netcoreapp2.1\jit-diff.dll
  jit-diff -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 23,77 ms for e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\jit-dasm.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\jit-dasm.csproj]
  jit-dasm -> e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm\bin\win10-x64\Release\netcoreapp2.1\jit-dasm.dll
  jit-dasm -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 25,93 ms for e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\jit-analyze.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core.
 See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\jit-analyze.csproj]
  jit-analyze -> e:\src\ms\dotnet\coreclr\jitutils\src\jit-analyze\bin\win10-x64\Release\netcoreapp2.1\jit-analyze.dll
  jit-analyze -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 23,47 ms for e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\jit-format.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\jit-format.csproj]
  jit-format -> e:\src\ms\dotnet\coreclr\jitutils\src\jit-format\bin\win10-x64\Release\netcoreapp2.1\jit-format.dll
  jit-format -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 24,25 ms for e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\cijobs.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\cijobs.csproj]
  cijobs -> e:\src\ms\dotnet\coreclr\jitutils\src\cijobs\bin\win10-x64\Release\netcoreapp2.1\cijobs.dll
  cijobs -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 16,69 ms for e:\src\ms\dotnet\coreclr\jitutils\src\pmi\pmi.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\pmi\pmi.csproj]
  pmi -> e:\src\ms\dotnet\coreclr\jitutils\src\pmi\bin\win10-x64\Release\netcoreapp2.1\pmi.dll
  pmi -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 20,56 ms for e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\jit-dasm-pmi.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\jit-dasm-pmi.csproj]
  jit-dasm-pmi -> e:\src\ms\dotnet\coreclr\jitutils\src\jit-dasm-pmi\bin\win10-x64\Release\netcoreapp2.1\jit-dasm-pmi.dll
  jit-dasm-pmi -> e:\src\ms\dotnet\coreclr\jitutils\bin\
        1 file(s) copied.
  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj...
  Restore completed in 4,53 sec for e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj.
Microsoft (R) Build Engine version 15.8.166+gd4e8d81a88 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restoring packages for e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj...
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\packages\obj\packages.csproj.nuget.g.props.
  Generating MSBuild file e:\src\ms\dotnet\coreclr\jitutils\src\packages\obj\packages.csproj.nuget.g.targets.
  Restore completed in 2,23 sec for e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj.
C:\Program Files\dotnet\sdk\3.0.100-alpha1-009460\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(143,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [e:\src\ms\dotnet\coreclr\jitutils\src\packages\packages.csproj]
  packages -> e:\src\ms\dotnet\coreclr\jitutils\src\packages\bin\win10-x64\Release\netcoreapp2.1\win10-x64\packages.dll
  packages -> e:\src\ms\dotnet\coreclr\jitutils\fx\
Downloading formatting tools
```

We can build jitutils repo and download formatting tools running `python -u tests\scripts\format.py -c e:\src\ms\dotnet\coreclr -o Windows_NT -a x64` will succeed as well.

Hope this explains the first group of problems which I have tried to fix.

File: bootstrap.cmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: Yes, the failures were due to version mismatch. 

https://github.com/dotnet/jitutils/blob/b74d29f4780f44d3d6bca689c931ddc6e84e7650/bootstrap.cmd#L132

File: bootstrap.cmd [hunk 5]
```diff
@@ -19,13 +19,13 @@ setlocal
...
-if not exist %__root%build.cmd goto clone_jitutils
+if not exist %__root%jitutils.sh goto clone_jitutils
 
 pushd %__root%
 
 REM Check if build.cmd is in the root of the repo.
-set __tempfile=%TEMP%\gittemp-%RANDOM%.txt
-git rev-parse --show-toplevel >%__tempfile% 2>&1
+set "__tempfile=%TEMP%\gittemp-%RANDOM%.txt"
+call git rev-parse --show-toplevel >"%__tempfile%"
```
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: Without call script fails to continue execution when run from command line

File: bootstrap.cmd [hunk 6]
```diff
@@ -122,19 +122,29 @@ GOTO CheckVersion
 
 :: Download clang-format and clang-tidy
 echo Downloading formatting tools
-call powershell Invoke-WebRequest -Uri "https://clrjit.blob.core.windows.net/clang-tools/windows/clang-format.exe" -OutFile bin\clang-format.exe
-call powershell Invoke-WebRequest -Uri "https://clrjit.blob.core.windows.net/clang-tools/windows/clang-tidy.exe" -OutFile bin\clang-tidy.exe
+
+call powershell -NoProfile Invoke-WebRequest -Uri "https://clrjit.blob.core.windows.net/clang-tools/windows/clang-format.exe" -OutFile bin\clang-format.exe
+if %errorlevel% NEQ 0 (
+    echo Failed to download clang-format
```
[the remark this answers is not in the document]
Comment (4creators) [reply bootstrap.cmd]: Addressed in new chengeset

**Comment by 4creators** [comment]
@BruceForstall @AndyAyersMS Addressed code review feedback. Added improved error handling. Provided explanation to why changes are needed. PTAL

**Comment by kunalspathak** [comment]
@BruceForstall @AndyAyersMS - Is the work in this PR still relevant? If yes, do you mind taking another look?

**Comment by BruceForstall** [comment]
Most of this was fixed with https://github.com/dotnet/jitutils/pull/224, so I'm going to close this.

If anyone is still experiencing problems or can suggest improvements, we could take an updated PR.

**Pull request closed without merging**